### PR TITLE
(Joe) plsql: persistence (WiP)

### PR DESCRIPTION
**Pull request by jsoroka**
almost not wip

### CODE REVIEW

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/DataDictionaryImpl.java [hunk 1]
```diff
@@ -8077,7 +8080,7 @@ public void grantPublicAccessToSystemRoutines(HashSet newlyCreatedRoutines,
         // Get schema ID for SYSCS_UTIL schema
         String schemaID=getSystemUtilSchemaDescriptor().getUUID().toString();
 
-        for(String routineName : sysUtilProceduresWithPublicAccess){
+        for(String routineName : sysUtilRoutinesWithPublicAccess){
```
Comment (shevek) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/DataDictionaryImpl.java]: Is this a bug fix?

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/DefaultSystemProcedureGenerator.java [hunk 2]
```diff
@@ -232,7 +232,7 @@ public final void createOrUpdateProcedure(
     	if (procedure == null) {
     		throw StandardException.newException(SQLState.LANG_OBJECT_NOT_FOUND_DURING_EXECUTION, "PROCEDURE", (schemaName + "." + procName));
     	} else {
-    		newlyCreatedRoutines.add(procedure.createSystemProcedure(schemaId, dictionary, tc));
+    		newlyCreatedRoutines.add(procedure.createSystemProcedure(schemaId, dictionary, tc).getMethodName());
```
Comment (shevek) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/DefaultSystemProcedureGenerator.java]: Is this a bug fix?

File: mem_sql/src/main/java/com/splicemachine/derby/lifecycle/MemDatabase.java [hunk 3]
```diff
@@ -87,11 +95,66 @@ public void run(){
...
+                    try { Thread.sleep(100); } catch (InterruptedException e) { ; }
+                }
+            } catch (Throwable t) {
+                partialShutdownException = t;
+                throw new IllegalStateException("server shutdown was interrupted with exception", t);
+            }
+        } else {
+            throw new IllegalStateException("attempt to shutdown server when it has not yet been started.");
+        }
+    }
```
Comment (shevek) [inline mem_sql/src/main/java/com/splicemachine/derby/lifecycle/MemDatabase.java]: I assume this is test integration.

File: pipeline_api/src/main/java/com/splicemachine/lifecycle/DatabaseLifecycleManager.java [hunk 4]
```diff
@@ -291,6 +291,7 @@ public void run(){
                     LOG.error("Error during shutdown of service "+ service+":",e);
                 }
             }
+            state.compareAndSet(State.SHUTTING_DOWN, State.SHUTDOWN);
```
Comment (shevek) [inline pipeline_api/src/main/java/com/splicemachine/lifecycle/DatabaseLifecycleManager.java]: Is this a bug fix?

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/BaseDataDictionary.java [hunk 5]
```diff
@@ -157,6 +157,7 @@
                                     "SYSCOLUMNSTATS",
                                     "SYSPHYSICALSTATS",
                                     "SYSTABLESTATS",
+                                    "SYSSOURCECODE",
```
Comment (shevek) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/BaseDataDictionary.java]: My only real comment: Should this be SYSPLSQLCODE ? What happens if we have to do other languages? Should we have another column?

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/BaseDataDictionary.java [hunk 6]
```diff
@@ -185,18 +186,13 @@
     };
 
 	/**
-	 * List of procedures in SYSCS_UTIL schema with PUBLIC access
+	 * List of procedures and functions in SYSCS_UTIL schema with PUBLIC access
 	 */
-	protected static final String[] sysUtilProceduresWithPublicAccess = { 
+	protected static final String[] sysUtilRoutinesWithPublicAccess = {
```
Comment (shevek) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/BaseDataDictionary.java]: While I may agree with this change, they may see it as unrelated.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/DefaultSystemProcedureGenerator.java (hunk 2 above)
Comment (jsoroka) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/DefaultSystemProcedureGenerator.java]: yes

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/BaseDataDictionary.java (hunk 5 above)
Comment (jsoroka) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/BaseDataDictionary.java]: I named it generically so as to not get Oracly PLSQL words into Splice opensource edition.  Furthermore, yeah, there's no reason why they can't store other languages in there, if they eventually support other stored program languages.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/DefaultSystemProcedureGenerator.java (hunk 2 above)
Comment (jsoroka) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/DefaultSystemProcedureGenerator.java]: Just a simple generics mistake.  It's a Set<String> and they're putting in ProcedureDefinitions or whatever the type is.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/BaseDataDictionary.java (hunk 6 above)
Comment (jsoroka) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/BaseDataDictionary.java]: It is, you're right.  Thanks.  I guess I was thinking about a /function/ that reads the source code and got wires crossed.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/DataDictionaryImpl.java (hunk 1 above)
Comment (jsoroka) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/DataDictionaryImpl.java]: yes, and an unrelated one, as you pointed out elsewhere in this PR.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/DefaultSystemProcedureGenerator.java (hunk 2 above)
Comment (jsoroka) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/DefaultSystemProcedureGenerator.java]: ... and it's not on purpose for any particular reason, just a mistake.

File: pipeline_api/src/main/java/com/splicemachine/lifecycle/DatabaseLifecycleManager.java (hunk 4 above)
Comment (jsoroka) [inline pipeline_api/src/main/java/com/splicemachine/lifecycle/DatabaseLifecycleManager.java]: Yes that is sort of a bug fix.  I suppose they don't need to actually set the State to SHUTDOWN in their normal operations, but when trying to shutdown an in-proc in-mem DB, we need to wait for that SHUTDOWN state to fire.
```
while (dlm.getState() != DatabaseLifecycleManager.State.SHUTDOWN
```

File: mem_sql/src/main/java/com/splicemachine/derby/lifecycle/MemDatabase.java (hunk 3 above)
Comment (jsoroka) [inline mem_sql/src/main/java/com/splicemachine/derby/lifecycle/MemDatabase.java]: Yes, right: PLSQLCompilerTest.

**Pull request closed without merging**